Tournament.cpp

```diff
@@ -0,0 +1,263 @@
+	}
+  }
+  participantsList.close();
+}
+void Tournament::setGroup(std::string& path, const int tournamentGroups) //The function distributes the participants into groups.
+{
+  std::ifstream groupList;
+  groupList.open(path);
+  if(!groupList.is_open() and !groupList.good())
+	throw std::exception("Unable to set groups. File not found!");
+  else
+  {
+	for (int numberGroups = 0; numberGroups < tournamentGroups; ++numberGroups)
+	{
+	  auto newGroup = std::make_shared<Group>();
+	  newGroup->teams.resize(newGroup->numberTeams);
+	  std::generate(newGroup->teams.begin(), newGroup->teams.end(), [&]()
+	  {
+		std::string name;
+		groupList >> name;
+		return addTeam(name);
+	  });
+	  m_groups.push_back(newGroup);
+	}
+  }
+  groupList.close();
+}
+std::shared_ptr<Team> Tournament::addTeam(std::string& name)
+{
+  for (auto it : m_participants)
+  {
+	if (it->name == name)
+	  return it;
+  }
+}
+void Tournament::playGroupRound()
+{
+  int numberRounds = 3;
+  for (int i = 0; i < numberRounds; ++i)
+  {
+	for (auto it : m_groups)
+	{
+	  int rank = 0;
+	  if (i == 0)
+	  {
+		std::cout << "----Round #1----" << std::endl;
+		it->playMatch(it->teams[0], it->teams[1]);
+		it->playMatch(it->teams[2], it->teams[3]);
+		std::cout << it<<std::endl;
+	  }
+	  if (i == 1)
+	  {
+		std::cout << "----Round #2----" << std::endl;
+		it->playMatch(it->teams[0], it->teams[2]);
+		it->playMatch(it->teams[1], it->teams[3]);
+		std::cout << it;
+	  }
+	  if (i == 2)
+	  {
+		std::cout << "----Round #3----" << std::endl;
+		it->playMatch(it->teams[0], it->teams[3]);
+		it->playMatch(it->teams[1], it->teams[2]);
+		it->sortGroup();
+		for_each(it->teams.begin(), it->teams.end(), [&](std::shared_ptr<Team>& team) {team->place = ++rank; });
+		std::cout << it;
+	  }
+	}
+  }
+}
+void Tournament::setEuroPlayOff()
+{
+  //We split the first, second and third places into three baskets.
+  std::queue<std::shared_ptr<Team>> firstTeams; //A1, B1, C1, D1, E1, F1
+  std::queue<std::shared_ptr<Team>> secondTeams; //A2, B2, C2, D2, E2, F2
+  std::vector<std::shared_ptr<Team>> preparationThirdTeams; //A3, B3, C3, D3, E3, F3
+  std::queue<std::shared_ptr<Team>> thirdTeams; //third-1, third-2, third-3, third-4
+  for (auto it : m_groups)
+  {
+	for_each(it->teams.begin(), it->teams.end(), [&](std::shared_ptr<Team>& team)
+	{
+	  if (team->place == 1)
+		  firstTeams.push(team);
+	  if (team->place == 2)
+		  secondTeams.push(team);
+	  if(team->place == 3)
+		  preparationThirdTeams.push_back(team);
+	});
+  }
+  //Find four best teams in third places
+  sort(preparationThirdTeams.begin(), preparationThirdTeams.end(), [](std::shared_ptr<Team>& lhs, std::shared_ptr<Team>& rhs)
+  {
+	return std::tie(lhs->points, lhs->goalsDifference) > std::tie(rhs->points, rhs->goalsDifference);
+  });
+  //Remove fifth and sixth teams
+  for (int i = 0; i < 2; ++i)
+  {
+	try
+	{
+	  if(preparationThirdTeams.empty())
+		throw std::exception("Vector preparationThirdTeams is empty");
```

Comment:
Why throw an exception here and wrap it in try-catch when you can just print the error here right away?